File: ts/src/builtins.ts
/* Copyright (c) 2026 Richard Rodger and other contributors, MIT License */

/*  builtins.ts
 *  Standard `$`-suffixed builtin function references.
 *
 *  A trailing `$` in a ref name marks an engine-provided builtin. The
 *  `$` ref-namespace is RESERVED: user grammars must not define refs
 *  containing `$` (enforced in `Tabnas.grammar()`). These functions are
 *  merged into the effective ref map at `grammar()` load time, so a
 *  *serialized* (function-free) GrammarSpec can reference them purely by
 *  name — the wire format carries strings like `@probeInit$`/`@node$`,
 *  not closures.
 *
 *  Two families:
 *
 *  1. Tree builders (`@node$`, `@capture$`, `@bubble$`) — rebuild the
 *     `{rule?, src, kids}` AST that the hand-written `@tabnas/abnf` plugin
 *     produces. Per-alt config rides in `alt.k.<name>` (the propagated
 *     `k` field), read from the action's 3rd argument so child rules are
 *     not polluted.
 *
 *  2. Probe + phase-retry dispatch (`@probeInit$`, `@probeDecide$`,
 *     `@probePhase0$/1$/2$`) — the optional-prefix `[X D] Y` ambiguity.
 *     The dispatcher rule retries itself across three phases stored in
 *     `r.k.pd_phase` (0 = probe, 1 = saw the disambiguator, 2 = did not).
 *     Phase 0 marks the position and runs a failure-proof probe; its
 *     close peeks the next token, rewinds, and commits to phase 1 or 2.
 *     The disambiguator token name rides in `k.pd_d` on the phase-0 open
 *     alt (propagated into `r.k`). These are generic — all grammar-
 *     specific data is config, never closed-over state.
 *
 *  CONTRACT: the AST shape and merge algorithm below MUST stay
 *  byte-identical to `@tabnas/abnf`'s `mkAstNode` / `segmentToAlt` /
 *  `captureChildFields`. The `BUILTIN_SCHEMA_VERSION` versions that
 *  contract; a serialized grammar may declare the schema it was compiled
 *  against and `grammar()` refuses a grammar that needs a newer one.
 */

import type { Rule, Context, AltMatch, AltAction, AltCond } from './types'


// The config-schema version implemented by these builtins. A serialized
// grammar that declares `GrammarSpec.v` greater than this is refused at
// load (see Tabnas.grammar). Absent ⇒ treated as version 1.
export const BUILTIN_SCHEMA_VERSION = 2


const defprop = Object.defineProperty

// Attach the engine's info marker as a hidden (non-enumerable) property on
// a node, so a grammar running with `info` on can introspect a container's
// origin (implicit flag, meta bag) or a string's quote without the marker
// leaking into JSON output. The TS info carrier (the marker property) is
// the counterpart of Go's MapRef/ListRef/Text wrapper structs.
function markNode(node: any, marker: string, data: any): void {
  if (node != null && 'object' === typeof node) {
    defprop(node, marker, { value: data, writable: true })
  }
}


// A builtin ref is either an alternate action (tree builders + probe
// actions) or an alternate condition (the phase guards).
export type BuiltinRef = AltAction | AltCond


// Output AST node produced by the tree builtins: `{rule?, src, kids}`.
// `user` rules carry a `rule` tag; `core`/`helper` rules omit it so they
// flatten into the enclosing user node.
interface AstNode {
  rule?: string
  src: string
  kids: any[]
}

function mkNode(rule: string | undefined, kind: string | undefined): AstNode {
  return 'user' === kind ? { rule, src: '', kids: [] } : { src: '', kids: [] }
}

// Per-alt tree-builder config, read from `alt.k.<name>`.
interface NodeConfig {
  init?: boolean
  rule?: string
  kind?: string
  nterms?: number
}
interface CaptureConfig {
  rule?: string
  kind?: string
}

// Per-alt config for the native-value builders. `slot` is the r.u key the
// pair key is stashed under (default 'key'); `from` is the matched-token
// index a key/scalar is read from (default 0). `implicit` (containers)
// records whether the container was created without delimiters — static
// config, not computed at close (see doc/value-builtins.md).
interface ObjectConfig {
  implicit?: boolean
}
interface ArrayConfig {
  implicit?: boolean
}
interface KeyConfig {
  slot?: string
  from?: number
}
interface SetvalConfig {
  slot?: string
}
interface ValueConfig {
  from?: number
}


// ---- Tree builders (config in `alt.k.<name>`) ---------------------
// These mirror @tabnas/abnf's emitter closures so a serialized,
// function-free grammar builds the identical `{rule,src,kids}` tree.

// Allocate (when `init`) and/or accumulate matched terminals' src.
const node$: AltAction = (r: Rule, _ctx: Context, alt: AltMatch) => {
  const cfg: NodeConfig = (alt && alt.k && alt.k.node$) || {}
  if (cfg.init) r.node = mkNode(cfg.rule, cfg.kind)
  const n = r.node as AstNode
  const nterms = cfg.nterms || 0
  for (let i = 0; i < nterms; i++) n.src += r.o[i].src
}

// Merge the just-returned child node into the current node. Tagged
// children push into `kids`; untagged ones flatten (src concatenates,
// kids extend).
const capture$: AltAction = (r: Rule, _ctx: Context, alt: AltMatch) => {
  const cfg: CaptureConfig = (alt && alt.k && alt.k.capture$) || {}
  if (null == r.node) r.node = mkNode(cfg.rule, cfg.kind)
  const n = r.node as AstNode
  const c = r.child && (r.child.node as any)
  if (null == c) return
  if ('object' !== typeof c || !('src' in c)) {
    n.kids.push(c)
    return
  }
  if (c === n) return
  n.src += c.src
  if (c.rule) n.kids.push(c)
  else if (Array.isArray(c.kids)) n.kids.push(...c.kids)
}

// Lift the committed child's node straight up (no merge).
const bubble$: AltAction = (r: Rule) => {
  if (r.child && (r.child.node as any) !== undefined) r.node = r.child.node
}


// ---- Probe + phase-retry dispatch ---------------------------------

// Phase-0 open action: mark the token position and reset phase.
const probeInit$: AltAction = (r: Rule, ctx: Context) => {
  r.k.pd_phase = 0
  r.k.pd_mark = ctx.mark()
}

// Phase-0 close action: peek the first token the probe did not consume,
// rewind to the mark, and commit to phase 1 (disambiguator present) or
// phase 2 (absent). The probe never fails, so `ctx.t[0]` always reflects
// a real position (the compiler emits a phase-0 close that consumes
// nothing — that contract is required for this builtin to be correct).
const probeDecide$: AltAction = (r: Rule, ctx: Context) => {
  // The phase-0 close must follow a phase-0 open (`@probeInit$`), which
  // records `pd_mark`. Guard against a malformed grammar that runs the
  // decision without it: `ctx.rewind(undefined)` would silently corrupt
  // the rewind window (NaN), breaking every later mark/rewind.
  if (null == r.k.pd_mark) {
    throw new Error(
      '@probeDecide$: no pd_mark — phase-0 @probeInit$ did not run')
  }
  const peek = ctx.t[0]
  ctx.rewind(r.k.pd_mark)
  r.k.pd_phase = peek && peek.name === r.k.pd_d ? 1 : 2
}

// Phase-guard conditions.
const probePhase0$: AltCond = (r: Rule) => !r.k.pd_phase
const probePhase1$: AltCond = (r: Rule) => 1 === r.k.pd_phase
const probePhase2$: AltCond = (r: Rule) => 2 === r.k.pd_phase


// ---- Native-value builders (config in `alt.k.<name>`) -------------
// Build NATIVE JSON values (objects/arrays/scalars), as opposed to the
// `{rule,src,kids}` syntax tree of the tree builders above. A grammar
// whose rules thread a node from parent to child (the engine seeds a
// pushed child's node from the parent — see makeRule) can assemble plain
// objects/arrays with these as alt actions. Schema family v2.
//
// These are INFO-AWARE: with `cfg.info` off they emit PLAIN nodes
// (byte-identical to v1); with `cfg.info.{map,list,text}` on they attach
// the engine's introspection marker (origin/quote metadata) — the same
// info model the Go engine carries via MapRef/ListRef/Text. The info
// logic lives here, in the engine, instead of each JSON-family plugin
// re-hand-writing it. See doc/value-builtins.md.

// Allocate a fresh empty object into r.node (no prototype, like JSON).
// When info.map is on, attach the marker with the static `implicit` flag.
const object$: AltAction = (r: Rule, ctx: Context, alt: AltMatch) => {
  const node = Object.create(null)
  r.node = node
  if (ctx.cfg.info.map) {
    const cfg: ObjectConfig = (alt && alt.k && alt.k.object$) || {}
    markNode(node, ctx.cfg.info.marker, { implicit: !!cfg.implicit, meta: {} })
  }
}

// Allocate a fresh empty array into r.node. When info.list is on, attach
// the marker with the static `implicit` flag.
const array$: AltAction = (r: Rule, ctx: Context, alt: AltMatch) => {
  const node: any[] = []
  r.node = node
  if (ctx.cfg.info.list) {
    const cfg: ArrayConfig = (alt && alt.k && alt.k.array$) || {}
    markNode(node, ctx.cfg.info.marker, { implicit: !!cfg.implicit, meta: {} })
  }
}

// Clear r.node back to "no value" — undoes the parent-seeded node so a
// scalar value (or fresh container) doesn't inherit the parent container.
const reset$: AltAction = (r: Rule) => {
  r.node = undefined
}

// Capture the matched key token's value into a (non-propagated) r.u slot,
// for a later @setval$ on the same rule to consume.
const key$: AltAction = (r: Rule, _ctx: Context, alt: AltMatch) => {
  const cfg: KeyConfig = (alt && alt.k && alt.k.key$) || {}
  r.u[cfg.slot || 'key'] = r.o[cfg.from || 0]?.val
}

// Assign the just-returned child node under the captured key: the object-
// property set. No-op if r.node isn't an object. When info.map is on, a
// key that collides with the marker is dropped (the marker rides as a
// hidden property, so a literal `"__info__"` key must not overwrite it) —
// this guard is TS-only; Go's field-based metadata has no key collision.
const setval$: AltAction = (r: Rule, ctx: Context, alt: AltMatch) => {
  const cfg: SetvalConfig = (alt && alt.k && alt.k.setval$) || {}
  const n = r.node as any
  if (null != n && 'object' === typeof n) {
    const key = r.u[cfg.slot || 'key']
    if (ctx.cfg.info.map && key === ctx.cfg.info.marker) return
    n[key] = r.child.node
  }
}

// Append the just-returned child node to the array (skips the no-value
// child). No-op if r.node isn't an array.
const push$: AltAction = (r: Rule) => {
  if (undefined !== r.child.node && Array.isArray(r.node)) {
    r.node.push(r.child.node)
  }
}

// Coalesce a value: a built child node wins; otherwise resolve the
// matched scalar token. When info.text is on and the resolved scalar is a
// string from a string/text token, box it as a `String` carrying the
// quote char (the leaf whose output type changes under info — it has no
// container to hang the marker on). The Go counterpart wraps in a `Text`.
const value$: AltAction = (r: Rule, ctx: Context, alt: AltMatch) => {
  if (undefined !== r.child.node) {
    r.node = r.child.node
    return
  }
  const cfg: ValueConfig = (alt && alt.k && alt.k.value$) || {}
  const tok = r.o[cfg.from || 0]
  let val = tok ? tok.resolveVal(r, ctx) : undefined
  const info = ctx.cfg.info
  if (
    info.text &&
    'string' === typeof val &&
    tok &&
    (tok.tin === ctx.cfg.t.ST || tok.tin === ctx.cfg.t.TX)
  ) {
    const quote = tok.tin === ctx.cfg.t.ST && tok.src.length > 0 ? tok.src[0] : ''
    const sv = new String(val)
    markNode(sv, info.marker, { quote })
    val = sv as any
  }
  r.node = val
}


// The standard builtin library, frozen so a grammar that (illegally)
// overrides a `$` ref cannot mutate the shared map for other instances.
export const BUILTIN_REFS: Readonly<Record<string, BuiltinRef>> = Object.freeze({
  '@node$': node$,
  '@capture$': capture$,
  '@bubble$': bubble$,
  '@probeInit$': probeInit$,
  '@probeDecide$': probeDecide$,
  '@probePhase0$': probePhase0$,
  '@probePhase1$': probePhase1$,
  '@probePhase2$': probePhase2$,

  // Native-value builders (schema v2).
  '@object$': object$,
  '@array$': array$,
  '@reset$': reset$,
  '@key$': key$,
  '@setval$': setval$,
  '@push$': push$,
  '@value$': value$,
})

File: rs/src/builtins.rs
// Copyright (c) 2013-2026 Richard Rodger, MIT License

use crate::options::InfoOptions;
use crate::rule::Rule;
use crate::token::{TIN_ST, TIN_TX};
use crate::value::{ListRef, MapRef, Text, Value};
use crate::Context;
use indexmap::IndexMap;
use std::cell::RefCell;
use std::rc::Rc;
use std::sync::Arc;

pub(crate) fn is_builtin_action(name: &str) -> bool {
    matches!(
        name,
        "@node$"
            | "@capture$"
            | "@bubble$"
            | "@fold$"
            | "@probeInit$"
            | "@probeDecide$"
            | "@object$"
            | "@array$"
            | "@reset$"
            | "@key$"
            | "@setval$"
            | "@push$"
            | "@value$"
            | "@map-bo"
            | "@list-bo"
            | "@pairkey"
            | "@pair-bc"
            | "@elem-bc"
            | "@val-bo"
            | "@val-bc"
    )
}

fn config_object(config: Option<&Value>) -> Option<&IndexMap<String, Value>> {
    config.and_then(|value| match value {
        Value::Object(map) => Some(map.as_ref()),
        _ => None,
    })
}

fn config_bool(config: Option<&Value>, name: &str) -> bool {
    matches!(
        config_object(config).and_then(|map| map.get(name)),
        Some(Value::Bool(true))
    )
}

fn config_string(config: Option<&Value>, name: &str) -> String {
    match config_object(config).and_then(|map| map.get(name)) {
        Some(Value::String(value)) => value.clone(),
        _ => String::new(),
    }
}

/// The config value at `name` only when it really is a string.
///
/// Distinct from `config_string`, which folds "absent" and "" into the
/// same empty String: `lit` needs them apart, because an empty string is
/// a real key. Matching the TS `typeof` guard and the Go type assertion,
/// anything that is not a string reads as absent, so the three ports
/// agree on every input and not merely on well-typed ones.
fn config_str_opt(config: Option<&Value>, name: &str) -> Option<String> {
    match config_object(config).and_then(|map| map.get(name)) {
        Some(Value::String(value)) => Some(value.clone()),
        _ => None,
    }
}

fn config_usize(config: Option<&Value>, name: &str) -> usize {
    match config_object(config).and_then(|map| map.get(name)) {
        Some(Value::Number(value)) if value.is_finite() && *value >= 0.0 => *value as usize,
        _ => 0,
    }
}

fn config_index(config: Option<&Value>, name: &str) -> Option<usize> {
    match config_object(config).and_then(|map| map.get(name)) {
        None => Some(0),
        Some(Value::Number(value))
            if value.is_finite()
                && value.fract() == 0.0
                && *value >= 0.0
                && *value <= usize::MAX as f64 =>
        {
            Some(*value as usize)
        }
        _ => None,
    }
}

fn ast_node(rule: String, kind: String) -> Value {
    let mut node = IndexMap::new();
    if kind == "user" {
        node.insert("rule".into(), Value::String(rule));
    }
    node.insert("src".into(), Value::String(String::new()));
    node.insert("kids".into(), Value::array(Vec::new()));
    Value::object(node)
}

/// The accumulated source text of a tree node -- the `{rule?, src, kids}`
/// shape `ast_node` builds -- or the node unchanged when it is not one.
///
/// "src" is how a member whose value IS its matched text gets that text:
/// the tree builders already accumulate it, and nothing else could read it
/// back out. A compiler emits "src" only where it already knows the member
/// is a scalar, so this never has to guess which it is: the fall-through
/// exists so asking for src where no tree node was built passes the value
/// along rather than erasing it.
fn src_val(node: Value) -> Value {
    if let Value::Object(map) = &node {
        if let Some(Value::String(source)) = map.get("src") {
            return Value::String(source.clone());
        }
    }
    node
}

fn append_src(node: &mut IndexMap<String, Value>, source: &str) {
    if let Some(Value::String(current)) = node.get_mut("src") {
        current.push_str(source);
    }
}

fn append_kid(node: &mut IndexMap<String, Value>, child: Value) {
    if let Some(kids) = node.get_mut("kids").and_then(Value::as_array_mut) {
        kids.push(child);
    }
}

fn capture_child(node: &mut IndexMap<String, Value>, child: Value) {
    if let Value::Object(child_map) = &child {
        if let Some(Value::String(source)) = child_map.get("src") {
            append_src(node, source);
            if child_map
                .get("rule")
                .is_some_and(|value| !matches!(value, Value::String(rule) if rule.is_empty()))
            {
                append_kid(node, child);
            } else if let Some(Value::Array(children)) = child_map.get("kids") {
                extend_kids(node, children);
            }
            return;
        }
    }
    append_kid(node, child);
}

/// Flatten an untagged child's kids into `node`'s.
///
/// A repetition compiles to untagged rules nested one per element, and
/// each level flattens the level below it, so a rule with N elements
/// flattens N + (N-1) + ... kids: quadratic in the element count of ONE
/// rule. TypeScript spreads the same way, but a JavaScript push is a
/// pointer store; here each kid cost a hash lookup, an `Arc` clone and a
/// later drop, and 800 fields took 46 seconds against a tenth of a
/// second (#195). A level whose own kids are still empty, which is every
/// level of that chain, now takes the child's array by handle instead:
/// one refcount bump, and the chain is linear again. Copy-on-write keeps
/// the child's own view intact should anything still read it.
fn extend_kids(node: &mut IndexMap<String, Value>, children: &Arc<Vec<Value>>) {
    match node.get_mut("kids") {
        Some(Value::Array(kids)) if kids.is_empty() => *kids = Arc::clone(children),
        Some(Value::Array(kids)) => Arc::make_mut(kids).extend(children.iter().cloned()),
        _ => {}
    }
}

fn map_value(info: &InfoOptions, implicit: bool) -> Value {
    if info.map {
        Value::MapRef(Arc::new(MapRef {
            value: IndexMap::new(),
            implicit,
            meta: IndexMap::new(),
        }))
    } else {
        Value::object(IndexMap::new())
    }
}

fn list_value(info: &InfoOptions, implicit: bool) -> Value {
    if info.list {
        Value::ListRef(Arc::new(ListRef {
            value: Vec::new(),
            implicit,
            child: None,
            meta: IndexMap::new(),
        }))
    } else {
        Value::array(Vec::new())
    }
}

fn token_value(rule: &mut Rule, context: &mut Context, index: usize, info: &InfoOptions) -> Value {
    let Some(token) = rule.o.get(index).cloned() else {
        return Value::Undefined;
    };
    let value = token.resolve_val(rule, context);
    if info.text && matches!(token.tin, TIN_ST | TIN_TX) {
        let string = match &value {
            Value::String(value) => value.clone(),
            Value::Text(value) => value.string.clone(),
            _ => return value,
        };
        let quote = if token.tin == TIN_ST {
            token
                .src
                .chars()
                .next()
                .map_or_else(String::new, |ch| ch.to_string())
        } else {
            String::new()
        };
        Value::Text(Text { quote, string })
    } else {
        value
    }
}

fn map_insert(node: &mut Value, key: String, value: Value) {
    match node {
        Value::Object(map) => {
            Arc::make_mut(map).insert(key, value);
        }
        Value::MapRef(map) => {
            Arc::make_mut(map).value.insert(key, value);
        }
        _ => {}
    }
}

fn list_push(node: &mut Value, value: Value) {
    match node {
        Value::Array(array) => Arc::make_mut(array).push(value),
        Value::ListRef(list) => Arc::make_mut(list).value.push(value),
        _ => {}
    }
}

/// Run a built-in with metadata wrappers disabled.
///
/// This keeps the original public helper stable for embedders. Parser-owned
/// execution uses [`run_builtin_action_with_info`] with its configured options.
pub fn run_builtin_action(name: &str, rule: &mut Rule, config: Option<&Value>) -> bool {
    let options = std::sync::Arc::new(crate::Options::default());
    let mut context = Context::new(
        options.rewind.history,
        "",
        Value::Undefined,
        std::sync::Arc::clone(&options),
        crate::InstanceInfo::default(),
    );
    run_builtin_action_with_info(name, rule, &mut context, config, &InfoOptions::default())
}

/// Run a builtin against an explicit live parse context. Embedders that use
/// lazy token values should prefer this over [`run_builtin_action`].
pub fn run_builtin_action_with_context(
    name: &str,
    rule: &mut Rule,
    context: &mut Context,
    config: Option<&Value>,
) -> bool {
    let info = context.options.info.clone();
    run_builtin_action_with_info(name, rule, context, config, &info)
}

pub(crate) fn run_builtin_action_with_info(
    name: &str,
    rule: &mut Rule,
    context: &mut Context,
    config: Option<&Value>,
    info: &InfoOptions,
) -> bool {
    match name {
        "@node$" => {
            if config_bool(config, "init") {
                rule.node = Rc::new(RefCell::new(ast_node(
                    config_string(config, "rule"),
                    config_string(config, "kind"),
                )));
            }
            let nterms = config_usize(config, "nterms");
            if let Some(node) = rule.node.borrow_mut().as_object_mut() {
                for token in rule.o.iter().take(nterms) {
                    append_src(node, &token.src);
                }
            }
        }
        "@capture$" => {
            if rule.node.borrow().is_undefined() {
                rule.node = Rc::new(RefCell::new(ast_node(
                    config_string(config, "rule"),
                    config_string(config, "kind"),
                )));
            }
            if !rule.child_node.is_undefined() && !rule.child_node_is_self {
                if let Some(node) = rule.node.borrow_mut().as_object_mut() {
                    capture_child(node, rule.child_node.clone());
                }
            }
        }
        "@bubble$" => {
            if rule.has_child_value() {
                rule.node = Rc::new(RefCell::new(rule.child_value()));
            }
        }
        "@fold$" => {
            if let Some(parent_node) = &rule.parent_node {
                let same_node = Rc::ptr_eq(parent_node, &rule.node);
                let own = rule.node.borrow().clone();
                if let Some(parent) = parent_node.borrow_mut().as_object_mut() {
                    if !same_node {
                        capture_child(parent, own);
                    }
                    let close_count = config_usize(config, "cN");
                    for token in rule.c.iter().take(close_count) {
                        append_src(parent, &token.src);
                    }
                }
            }
            rule.node = Rc::new(RefCell::new(Value::Undefined));
        }
        "@val-bo" => {
            rule.node = Rc::new(RefCell::new(Value::Undefined));
        }
        "@val-bc" => {
            let is_undef = rule.node.borrow().is_undefined();
            if is_undef {
                if rule.has_child_value() {
                    rule.node = Rc::new(RefCell::new(rule.child_value()));
                } else if rule.os() > 0 {
                    rule.node = Rc::new(RefCell::new(token_value(rule, context, 0, info)));
                }
            }
        }
        "@value$" => {
            if rule.has_child_value() {
                rule.node = Rc::new(RefCell::new(rule.child_value()));
            } else {
                let value = config_index(config, "from").map_or(Value::Undefined, |index| {
                    token_value(rule, context, index, info)
                });
                rule.node = Rc::new(RefCell::new(value));
            }
        }
        "@object$" => {
            rule.node = Rc::new(RefCell::new(map_value(
                info,
                config_bool(config, "implicit"),
            )));
        }
        "@array$" => {
            rule.node = Rc::new(RefCell::new(list_value(
                info,
                config_bool(config, "implicit"),
            )));
        }
        "@reset$" => {
            rule.node = Rc::new(RefCell::new(Value::Undefined));
            rule.child_node = Value::Undefined;
            rule.child_node_is_self = false;
        }
        "@key$" => {
            let slot = {
                let configured = config_string(config, "slot");
                if configured.is_empty() {
                    "key".to_string()
                } else {
                    configured
                }
            };
            // `lit` supplies the key as a CONSTANT instead of reading it
            // from a token, and wins outright when set. A grammar whose
            // structure is declared rather than delimited -- `ver = maj
            // "." min`, where `maj` names a part but no token carries the
            // text "maj" -- has no token for @key$ to read, so without
            // this the key side of @setval$ is unreachable for it.
            if let Some(lit) = config_str_opt(config, "lit") {
                rule.u_mut().insert(slot, Value::String(lit));
            } else if let Some(value) = config_index(config, "from")
                .and_then(|index| rule.o.get(index))
                .map(|token| token.val.clone())
            {
                rule.u_mut().insert(slot, value);
            }
        }
        "@setval$" => {
            let slot = {
                let configured = config_string(config, "slot");
                if configured.is_empty() {
                    "key".to_string()
                } else {
                    configured
                }
            };
            if let Some(Value::String(key)) = rule.u.get(&slot).cloned() {
                let mut val = rule.child_value();
                if config_bool(config, "src") {
                    val = src_val(val);
                }
                map_insert(&mut rule.node.borrow_mut(), key, val);
            }
        }
        "@push$" => {
            if rule.has_child_value() {
                let mut val = rule.child_value();
                if config_bool(config, "src") {
                    val = src_val(val);
                }
                list_push(&mut rule.node.borrow_mut(), val);
            }
        }
        "@map-bo" => {
            rule.node = Rc::new(RefCell::new(map_value(info, false)));
        }
        "@list-bo" => {
            rule.node = Rc::new(RefCell::new(list_value(info, false)));
        }
        // The captured key belongs only to this pair rule and is consumed by
        // @pair-bc on that same rule. Store it in `u`, the non-propagating
        // scratch bag, so it cannot leak into child rules as `n` or `k` would.
        "@pairkey" => {
            if let Some(t0) = rule.o0() {
                let key = match &t0.val {
                    Value::String(s) => s.clone(),
                    _ => t0.src.to_string(),
                };
                rule.u_mut().insert("key".to_string(), Value::String(key));
            }
        }
        "@pair-bc" => {
            if let Some(Value::String(key)) = rule.u.get("key").cloned() {
                let child = rule.child_value();
                let mut node = rule.node.borrow_mut();
                map_insert(&mut node, key, child);
            }
        }
        "@elem-bc" if rule.has_child_value() => {
            let child = rule.child_value();
            let mut node = rule.node.borrow_mut();
            list_push(&mut node, child);
        }
        _ => return false,
    }
    true
}
